dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/CheckInUnAssignWorkflowStepCheckerListener.java

```diff
@@ -0,0 +1,141 @@
+     * @param customAutoAssignWorkflowDelegate
+     */
+    public void setCustomAutoAssignWorkflowDelegate (final AutoAssignWorkflowDelegate customAutoAssignWorkflowDelegate) {
+
+        if (null != customAutoAssignWorkflowDelegate) {
+
+            this.customAutoAssignWorkflowDelegate = customAutoAssignWorkflowDelegate;
+        }
+    }
+
+    @Override
+    public void notify(final ContentletCheckinEvent event) {
+
+        try {
+
+            if (!APILocator.getWorkflowAPI().findCurrentStep(event.getContentlet()).isPresent()) {
+
+                if (Config.getBooleanProperty("AUTO_ASSIGN_WORKFLOW", true)) {
+
+                    final AutoAssignWorkflowDelegate delegate = null != this.customAutoAssignWorkflowDelegate?
+                            this.customAutoAssignWorkflowDelegate: this.autoAssignWorkflowDelegate;
+                    Logger.debug(this, "Using the auto assign workflow with the delegate: " + delegate.getName());
+
+                    delegate.assign(event.getContentlet(), event.getUser());
+                }
+            }
+        } catch (DotDataException e) {
+
+            Logger.error(this, e.getMessage(), e);
+        }
+    } // notify.
+
+    private static void assign (final Contentlet contentlet, final User user) {
+
+        List<WorkflowAction> workflowActions = null;
+
+        try {
+
+            workflowActions = APILocator.getWorkflowAPI()
+                    .findAvailableDefaultActionsByContentType(contentlet.getContentType(), user);
+
+            if (UtilMethods.isSet(workflowActions) && workflowActions.size() > 0) {
+
+                APILocator.getWorkflowAPI().saveWorkflowTask(
+                        createWorkflowTask(contentlet, user, workflowActions));
+            } else {
+
+                setToFirstSystemWorkflowStep (contentlet, user);
+            }
+        } catch (DotDataException | DotSecurityException e) {
+
+            Logger.error(CheckInUnAssignWorkflowStepCheckerListener.class, e.getMessage(), e);
+        }
+    }
+
+    private static WorkflowTask createWorkflowTask(final Contentlet contentlet,
+                                                   final User user,
+                                                   final List<WorkflowAction> workflowActions) throws DotDataException {
+
+        final WorkflowAction firstWorkflowAction    = workflowActions.get(0);
+        final WorkflowAction selectedWorkflowAction = workflowActions.stream()
+                .filter(workflowAction -> workflowAction.getSchemeId().equals(SystemWorkflowConstants.SYSTEM_WORKFLOW_ID))
+                .findFirst().orElse(firstWorkflowAction);
+        final WorkflowStep   workflowStep           = APILocator.getWorkflowAPI().findFirstStepForAction(selectedWorkflowAction).get();
+        return createWorkflowTask(contentlet, user, workflowStep);
+    }
+
+    private static WorkflowTask createWorkflowTask(final Contentlet contentlet, final User user,
```

Comment:
This should be in the workflowAPI with a test, right?

Reply:
OK I will

Reply:
Done